cilium_eip_no_masquerade_agent/src/main.rs

```diff
@@ -98,36 +81,125 @@ impl k8s_controller::Context for Context {
+            .map(|(i, _)| i);
+        if let Some(finalizer_index) = finalizer_index {
+            let pod_api: Api<Pod> =
+                Api::namespaced(self.kube_client.clone(), &pod.namespace().unwrap());
+            let finalizer_path = format!("/metadata/finalizers/{finalizer_index}");
+            pod_api
+                .patch::<Pod>(
+                    pod_name,
+                    &PatchParams::default(),
+                    &Patch::Json(json_patch::Patch(vec![
+                        // All finalizers run concurrently and we use an integer index
+                        // `Test` ensures that we fail instead of deleting someone else's finalizer
+                        PatchOperation::Test(TestOperation {
+                            path: finalizer_path.clone(),
+                            value: FINALIZER_NAME.into(),
+                        }),
+                        PatchOperation::Remove(RemoveOperation {
+                            path: finalizer_path,
+                        }),
+                    ])),
+                )
+                .await
+                .unwrap();
+        }
+    }
+
+    async fn wait_for_chain_to_exist(&self) -> Result<(), Box<dyn std::error::Error>> {
+        info!("Waiting for {CHAIN} chain to exist");
+        while !self.iptables.chain_exists(TABLE, CHAIN)? {
+            trace!("{CHAIN} does not exist, sleeping");
+            sleep(Duration::from_secs(1)).await;
+        }
+        Ok(())
+    }
+
+    async fn get_eni_indexes_for_existing_ip_rules(
+        &self,
+    ) -> Result<HashSet<u32>, rtnetlink::Error> {
+        let rules = self.ip_rule_handle.get(IpVersion::V4).execute();
+        rules
+            .try_filter(|rule| future::ready(rule.nlas.contains(&Nla::FwMask(0xf))))
+            .map_ok(|rule| {
+                rule.nlas
+                    .into_iter()
+                    .find_map(|nla| match nla {
+                        Nla::FwMark(i) => Some(i),
+                        _ => None,
+                    })
+                    .expect("There should always be an associated mark if there is a mask.")
+            })
+            .collect::<Vec<Result<u32, rtnetlink::Error>>>()
+            .await
+            .into_iter()
+            .collect()
+    }
+
+    fn ensure_iptables_rule(&self, rule: &str) -> Result<(), Box<dyn std::error::Error>> {
+        if !self.iptables.exists(TABLE, CHAIN, rule)? {
+            event!(Level::INFO, rule = ?rule, "Appending iptables rule");
+            self.iptables.append(TABLE, CHAIN, rule)?;
+        }
+        Ok(())
+    }
+
+    fn ensure_restore_mark_iptables_rule(&self) -> Result<(), Box<dyn std::error::Error>> {
+        let rule = "-i lxc+ -m comment --comment \"cilium: secondary interfaces\" -j CONNMARK --restore-mark --nfmask 0xf --ctmask 0xf";
+        self.ensure_iptables_rule(rule)?;
+        Ok(())
+    }
+
+    fn ensure_mark_iptables_rule(
+        &self,
```

Comment:
supernit: it's odd to me that the iptables library allows this, but I'd think we'd want to borrow `&mut self` for these fns that mutate the iptable state so it's always given exclusive access

Reply:
I assume it's because it isn't actually mutating any rust objects. All the state is in the kernel. I can switch to &mut self.